src/permutation.jl

```diff
@@ -0,0 +1,162 @@
+    m_threads = [m]
+    βsc_threads = [βsc]
+    θsc_threads = [θsc]
+
+    if use_threads
+        Threads.resize_nthreads!(m_threads)
+        Threads.resize_nthreads!(βsc_threads)
+        Threads.resize_nthreads!(θsc_threads)
+    end
+    # we use locks to guarantee thread-safety, but there might be better ways to do this for some RNGs
+    # see https://docs.julialang.org/en/v1.3/manual/parallel-computing/#Side-effects-and-mutable-function-arguments-1
+    # see https://docs.julialang.org/en/v1/stdlib/Future/index.html
+    rnglock = ReentrantLock()
+    samp = replicate(n, use_threads=use_threads) do
+        mod = m_threads[Threads.threadid()]
+
+        local βsc = βsc_threads[Threads.threadid()]
+        local θsc = θsc_threads[Threads.threadid()]
+        lock(rnglock)
+        mod = permute!(rng, mod, blups, reterms)
+        unlock(rnglock)
+        refit!(mod)
+        (
+         objective = mod.objective,
+         σ = mod.σ,
+         β = NamedTuple{β_names}(fixef!(βsc, mod)),
+         se = SVector{p,T}(stderror!(βsc, mod)),
+         θ = SVector{k,T}(getθ!(θsc, mod)),
+        )
+    end
+    MixedModelPermutation(
+        samp,
+        deepcopy(morig.λ),
+        getfield.(morig.reterms, :inds),
+        copy(morig.optsum.lowerbd),
+        NamedTuple{Symbol.(fnames(morig))}(map(t -> (t.cnames...,), morig.reterms)),
+    )
+end
+
+function permutation(nsamp::Integer, m::LinearMixedModel; kwargs...)
+    return permutation(Random.GLOBAL_RNG, nsamp, m; kwargs...)
+end
+
+function permutation(rng::AbstractRNG, n::Integer,
+                                morig::GeneralizedLinearMixedModel;
+                                use_threads::Bool=false) where {T}
+
+    throw(ArgumentError("GLMM support is not yet implemented"))
+end
+
+permute!(mod::LinearMixedModel, blups=ranef(mod), reterms=mod.reterms) =
+    permute!(GLOBAL_RNG, mod, blups, reterms)
+
+"""
+    permute!([rng::AbstractRNG,] mod::LinearMixedModel,
+              blups=ranef(mod), reterms=mod.reterms)
+
+Simulate and install a new response via sign-flipped permutation of the residuals
+at the observational and sign-flipping of the conditional modes at group level.
+
+Sign-flipped permutation of the residuals is similar to permuting the
+(fixed-effects) design matrix. Sign-flipping the conditional modes (random effects)
+preserves the correlation structure of the random effects, while also being equivalent to
+permutation via swapped labels for categorical variables.
+
+!!! note
+    This method has serious limitations for singular models because permuting zeros
+
+See also [`permutation`](@ref), [`nonparametricbootstrap`](@ref) and [`resample!`](@ref).
+
+# Reference
+ter Braak?
+
+"""
+function permute!(rng::AbstractRNG, mod::LinearMixedModel,
+                   blups=ranef(mod),
+                   reterms=mod.reterms)
+    β = coef(mod)
+    y = response(mod) # we are now modifying the model
+    copy!(y, residuals(mod))
+
+    shuffle!(rng, y)
```

Comment:
Residuals are permuted here

Reply:
I am not sure where is the effect of "shuffle()". It looks like it modify the "y" object (= the response variable) and not the residuals.

Reply:
Sorry I fergot to read row 125.